contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1047 @@
+        schedule.tokenHolders.add(to);
+      }
+      if (balanceOf(from, ids[i]) == 0) {
+        // slither-disable-next-line unused-return return value irrelevant, id guaranteed removed
+        _addressToScheduleIdSet[from].remove(ids[i]);
+        // slither-disable-next-line unused-return return value irrelevant, address guaranteed removed
+        schedule.tokenHolders.remove(from);
+      }
+    }
+  }
+
+  /**
+   * For each index-matched account, removal id, and amount, revokes amount of tokens from schedule corresponding
+   * to the specified removal and transfers to the corresponding account.
+   *
+   * @dev Transfers any unreleased tokens in the removal id's corresponding schedule and reduces the total supply
+   * of that token. No change is made to balances that have released but not yet been claimed.
+   * If a token has multiple owners, balances are burned proportionally to ownership percentage, summing to the total
+   * amount being revoked.
+   *
+   * The behavior of this function can be used in two specific ways:
+   * - To revoke a specific number of tokens as specified by the `amounts` array.
+   * - To revoke all remaining revokable tokens in a schedule set amount to 0 in the `amounts` array.
+   *
+   * ##### Requirements:
+   *
+   * - Can only be used when the caller has the `TOKEN_REVOKER_ROLE`
+   * - The requirements of _beforeTokenTransfer apply to this function
+   * - toAccounts.length == removalIds.length == amounts.length
+   */
+  function batchRevokeUnreleasedTokenAmounts(
+    address[] calldata toAccounts,
+    uint256[] calldata removalIds,
+    uint256[] calldata amounts
+  ) external whenNotPaused onlyRole(TOKEN_REVOKER_ROLE) {
+    if (!(toAccounts.length == removalIds.length)) {
+      revert ArrayLengthMismatch({
+        array1Name: "fromAccounts",
+        array2Name: "removalIds"
+      });
+    }
+    if (!(toAccounts.length == amounts.length)) {
+      revert ArrayLengthMismatch({
+        array1Name: "fromAccounts",
+        array2Name: "amounts"
+      });
+    }
+    for (uint256 i = 0; i < toAccounts.length; i++) {
+      _revokeUnreleasedTokens(toAccounts[i], removalIds[i], amounts[i]);
+    }
+  }
+
+  // Private implementations ==========================================
+
+  /**
+   * Wraps minting of wrapper token and schedule setup.
+   *
+   * @dev If no schedule is set up for the specified removal id, one is created.
+   *
+   * @param removalId uint256 The removal for which funds are being deposited.
+   * @param amount uint256 Quantity of `_bridgedPolygonNori` to deposit
+   * @param userData bytes extra information provided by the user (if any)
+   * @param operatorData bytes extra information provided by the operator (if any)
+   */
+  function _depositFor(
+    uint256 removalId,
+    uint256 amount,
+    bytes memory userData,
+    bytes memory operatorData
+  ) internal returns (bool) {
+    uint256 projectId = _removal.getProjectIdForRemoval(removalId);
```

Comment:
yet another reason to find a way to get projectId into the remoavl id...

Reply:
would definitely be nice... I assume you saw the thread though where we realized we really can't remove the parcel id (`subIdentifier`) from the id?  for the same reason, we require the vintage.  so that doesn't leave much wiggle room unless we decide to get rid of the supplier address as part of the removal id and replace it with the project id.  It seems like the supplier address is referenced more frequently from the removal id so there's a case for prioritizing that over the project id but maybe worth revisiting, especially if we ever want to create some indirection around the relationship between removal ids and supplier addresses so that addresses to be paid can be updated more easily in the case of a compromise or something

Reply:
Yeah, I mentioned removing the supplier address instead [here](https://github.com/nori-dot-eco/contracts/pull/249/files#r903174741) because encoding that is problematic anyways

Reply:
a todo or ticket for now would suffice. I think it's dangerous to encode the supplier address considering it cannot be fully trusted